### PR TITLE
62244 - Properly Connect and Gate the Acceptable Verified Credential Adoption Service

### PR DESCRIPTION
**Pull request by jrjamespdx**
## Summary

**Changes Made:**
This PR does the following:
- Changes the name of the already present `reactivation_experiment` Flipper toggle to `reactivation_experiment_rate_limiting` in `app/services/acceptable_verified_credential_adoption_service.rb`
- Adds a new Flipper toggle to `app/services/login/after_login_actions.rb` called `reactivation_experiment_initial_gate` that will control the basic "on/off" for users's being checked for eligibility to be sent a credential adoption related email.
- Updates the `config/features.yml` to align with the previous two points
- Updates the tests for both services to reflect the new and modified feature toggles

**What do these changes solve/make possible?**
These changes enable us to both gate and rate limit the number of users going through the AcceptableVerifiedCredentialAdoptionService flow.  

**Why are you using two Flipper toggles?**
- We only want user eligibility checked for the time we have the experiment "running" (outlined below).
- We want to restrict the total number of emails being sent out to 5% of daily logged in users who also meet the other eligibility criteria (outlined further below).

More specifically:
- The initial gate feature toggle will give us the ability to easily turn the entire feature on or off to prevent the service from being called and evaluating other eligibility criteria after every time a user logs in.
- The rate limit feature toggle will give us the ability to restrict the percentage of users who meet the other eligibility criteria from receiving the email during this experiment.

Hypothetical Example with easy numbers:
During the experiment window, if 1000 users login daily we want them all to go through the eligibility check within the service (the initial gate toggle).  If 100 users of the original 1000 are eligible, we only want a certain percentage of those folks to actually be sent the email (the rate limit toggle).  So if our rate limit Flipper is set to 5% of logged on users, we'd anticipate 5 users would go through the remainder of the logic to be sent the email (barring any other confounding factors).

Note:  Currently, an eligible user is a veteran user who meets the following criteria: 
- user has both a legacy (DSLogon) and a modern (Login.gov) credential 
- user has most recently logged in with the legacy credential
- user has not triggered a previous Credential Adoption email send in the last seven (7) days

**Team Info:**
This work is being done by the Login.gov Adoption Team which reports to John Rahaghi, Product Owner with OCTO on the VA side.

**Flipper Info:**
`reactivation_experiment_initial_gate` and `reactivation_experiment_rate_limit`
- this toggle will be enabled for a maximum of one (1) week, starting on/around 20JULY2023 depending on deployment timing
- this toggle will be turned off once the one (1) week max time limit is hit OR if we reach 5% of our eligible population, which ever happens first
- success criteria is based on the percentage of our total eligible veteran user base

## Related issue(s)
- [Connect Credential Adoption Service for Production Usage#62244](https://app.zenhub.com/workspaces/logingov-adoption-team-632280b31e745000136a61fc/issues/gh/department-of-veterans-affairs/va.gov-team/62244)

## Testing done

**Behavior Before Change**
Previously, there was no call to the AcceptableVerifiedCredentialAdoptionService from within the AfterLoginActions service so it was unused code.  

**Verification Steps**
As the VA Notify portion of the flow does not work in a local environment, the process for verifying the behavior was as follows:
- create local users with mocked credentials and attributes that would enable them to be considered eligible
- create local users with mocked credentials and attributes that would NOT enable them to be considered eligible
- temporarily adjust logging methods to log to the console versus StatsD
- simulate logging in with each user, observing that the appropriate log entries and/or database records were or were not created based on if the user was actually eligible or not.

This method of verification was very time consuming and required several mocks and temp changes to logging methods due to our current inability to "pre-test/QA" in a staging environment.  As such I do not recommend it.

The tests for both the AfterLoginAction service and the AcceptableVerifiedCredentialAdoptionService test these behaviors as well, albeit in a limited piecemeal fashion.

## What areas of the site does it impact?
This PR makes no changes to the UI of the site itself.  It enables access to a previously created and tested code path, with that access being controlled by the new/update Flipper toggles.

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Additional Notes

This feature will be short-lived and will be removed once the Login.gov Adoption Team has fully transitioned to a new GovDelivery instance.  As such, we're trying to keep this service very simple and decoupled as possible.

### CODE REVIEW

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (holytoastr) [inline app/services/acceptable_verified_credential_adoption_service.rb:103]: Good clarification comment for future us.

File: app/services/login/after_login_actions.rb
Comment (holytoastr) [inline app/services/login/after_login_actions.rb:24]: Could probably tighten this up as:

`AcceptableVerifiedCredentialAdoptionService.new(@current_user).perform if Flipper.enabled?(:reactivation_experiment_initial_gate, current_user)`

File: config/features.yml
Comment (holytoastr) [inline config/features.yml:508]: The flipper flag clean up task sometimes doesn't work, so you might need to log into the console and run the following to remove the old feature name:
`Flipper.remove("name_of_feature_flag")`

File: app/services/login/after_login_actions.rb
Comment (holytoastr) [inline app/services/login/after_login_actions.rb:25]: Also, I don't think we need the `@` on `current_user` here.

**Review by holytoastr — COMMENTED** [review]
Looks really good, but needs a little clean up! Thank you for tackling this!

File: app/services/login/after_login_actions.rb
Comment (jrjamespdx) [reply app/services/login/after_login_actions.rb:24]: Changed!

File: app/services/login/after_login_actions.rb
Comment (jrjamespdx) [reply app/services/login/after_login_actions.rb:25]: Changed!

File: config/features.yml
Comment (jrjamespdx) [reply config/features.yml:508]: Gotcha, I'll double check it once this gets merged and deployed

**Comment by jrjamespdx** [comment]
For whatever reason my pushed commits are not updating this PR.

**Review by holytoastr — APPROVED** [review]
Looks good to move to the next stage of reviews! Thanks for making those changes!

**Review by ryan-mcneil — APPROVED** [review]
👍

**Comment by bosawt** [comment]
Wait, why was this merged without an identity team approval?

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (bosawt) [inline app/services/acceptable_verified_credential_adoption_service.rb:95]: Weren't you going to remove the check for 7 days ago in this PR?

**Comment by jrjamespdx** [comment]
> @jrjamespdx Wait, why was this merged without an identity team approval?

@bosawt 
This was 100% on me moving too fast after today's craziness.  We got the notification that there were approvals and when I went to see if it was approved to merge it said "Approved and Ready to Merge" with "2 Reviewers".  I mistakenly assumed that if it was giving me the ok to merge that I had all the required approvals from the all required code owners.

I'll reach out to you on the VFS Slack to discuss potentially reverting this until I have your approval as well.

**Comment by jrjamespdx** [comment]
Update for the sake of posterity:   I spoke with @bosawt and we've decided to not revert this PR.  Trevor also identified that the related files weren't part of the Identity Team code owners, so that will be addressed to ensure that their reviews are also required when necessary moving forward.

**Comment by bosawt** [comment]
> Update for the sake of posterity: I spoke with @bosawt and we've decided to not revert this PR. Trevor also identified that the related files weren't part of the Identity Team code owners, so that will be addressed to ensure that their reviews are also required when necessary moving forward.

This is correct. I just ask for suitable testing on staging tonight/tomorrow before the production deploy to make up for lack of dev acceptance from Identity team. Metrics to watch out for:
- make sure auth rates are not affected
- make sure enabling percentage of users through flipper gate results in expected numbers of users being sent emails
- ensure only users who are eligible for emails in the first place are being invoked 
I think if these are shown then there's a high confidence this will not negatively affect production

**Pull request merged**
